sdk/src/wallet/core/builder.rs

```diff
@@ -270,6 +270,9 @@ where
         // If the wallet builder is not set, it means the user provided it and we need to update the addresses.
         // In the other case it was loaded from the database and addresses are up to date.
         if provided_client_options {
+            // Can't call it in wasm, because it will panic with `condvar wait not supported`. Updating the bech32 hrp
```

Comment:
Could you explain this more?

Reply:
It's not possible to block on awaits that need to interact with the JS side in wasm in a sync function
The function is calling the nodeinfo and then it panics
Full error is
```
panicked at library/std/src/sys/wasm/../unsupported/locks/condvar.rs:20:9:
condvar wait not supported

Stack:

Error
    at module.exports.__wbg_new_abda76e883ba8a5f (/home/t/Desktop/iota-sdk/bindings/wasm/node/wasm/iota_sdk_wasm.js:889:17)
    at wasm://wasm/0164e37e:wasm-function[1198]:0x483274
    at wasm://wasm/0164e37e:wasm-function[2587]:0x50c588
    at wasm://wasm/0164e37e:wasm-function[2777]:0x515608
    at wasm://wasm/0164e37e:wasm-function[3887]:0x526a90
    at wasm://wasm/0164e37e:wasm-function[3203]:0x521228
    at wasm://wasm/0164e37e:wasm-function[3236]:0x521abc
    at wasm://wasm/0164e37e:wasm-function[3241]:0x521bef
    at wasm://wasm/0164e37e:wasm-function[2078]:0x4ecce2
    at wasm://wasm/0164e37e:wasm-function[1084]:0x46b789


wasm://wasm/0164e37e:1


RuntimeError: unreachable
    at wasm://wasm/0164e37e:wasm-function[3896]:0x526ad5
    at wasm://wasm/0164e37e:wasm-function[2587]:0x50c5b1
    at wasm://wasm/0164e37e:wasm-function[2777]:0x515608
    at wasm://wasm/0164e37e:wasm-function[3887]:0x526a90
    at wasm://wasm/0164e37e:wasm-function[3203]:0x521228
    at wasm://wasm/0164e37e:wasm-function[3236]:0x521abc
    at wasm://wasm/0164e37e:wasm-function[3241]:0x521bef
    at wasm://wasm/0164e37e:wasm-function[2078]:0x4ecce2
    at wasm://wasm/0164e37e:wasm-function[1084]:0x46b789
    at wasm://wasm/0164e37e:wasm-function[120]:0x1b8e07
```

Reply:
Sync function because we're under `createWallet`?

Reply:
Correct, in the JS constructors we can't be async

Reply:
Maybe we shouldn't use constructors for this then, this is extremely limiting and hidden bugs are clearly possible.

Reply:
So how do you create a wallet then if not with a constructor?

Reply:
Use a constructor that does nothing and then call a method to fill in the data, I guess?

Reply:
Like this? Bit ugly
```JS
class WalletOptions { }
class WalletMethodHandler { }
class Wallet {
    methodHandler: WalletMethodHandler
    options: WalletOptions

    constructor(options: WalletOptions) {
        this.options = options;
    }

    async finish(): Promise<Wallet> {
        this.methodHandler = await createWallet(this.options);
        return this
    }
}

async function createWallet(options: WalletOptions): Promise<WalletMethodHandler> {
    return new WalletMethodHandler()
}

async function main() {
    let wallet = await new Wallet({}).finish();
}
main()
```

Reply:
I would go for a static async constructor instead:
```js
class WalletOptions { }
class WalletMethodHandler { }
class Wallet {
    methodHandler: WalletMethodHandler
    options: WalletOptions

    constructor(options: WalletOptions, methodHandler: WalletMethodHandler) {
        this.options = options;
        this.methodHandler = methodHandler;
    }

    static async create(options: WalletOptions): Promise<WalletMethodHandler> {
        return new Wallet(options, new WalletMethodHandler())
    }
}

async function main() {
    let wallet = await Wallet.create({});
}
```